spec/filters/geoip_ecs_spec.rb

```diff
@@ -0,0 +1,203 @@
+        end
+
+        context "`target` is unset" do
+
+          context "`source` end with [ip]" do
+            let(:event) { LogStash::Event.new("host" => {"ip" => ip}) }
+            let(:options) { common_options.merge({"source" => "[host][ip]"}) }
+
+            it "should use [host] as target" do
+              plugin.register
+              plugin.filter(event)
+              expect( event.get "[host][geo][country_iso_code]" ).to eq 'US'
+            end
+          end
+
+          context "`source` end with [ip] but `target` does not match ECS template" do
+            let(:event) { LogStash::Event.new("hostname" => {"ip" => ip}) }
+            let(:options) { common_options.merge({"source" => "[hostname][ip]"}) }
+
+            it "should use [hostname] as target with warning" do
+              expect(plugin.logger).to receive(:warn)
+              plugin.register
+              plugin.filter(event)
+              expect( event.get "[hostname][geo][country_iso_code]" ).to eq 'US'
+            end
+          end
+
+          context "`source` == [ip]" do
+            let(:event) { LogStash::Event.new("ip" => ip) }
+            let(:options) { common_options.merge({"source" => "[ip]"}) }
+
+            it "should raise error to require `target`" do
+              expect { plugin.register }.to raise_error LogStash::ConfigurationError
+            end
+          end
+
+          context "`source` not end with [ip]" do
+            let(:event) { LogStash::Event.new("host_ip" => ip) }
+            let(:options) { common_options.merge({"source" => "host_ip"}) }
+
+            it "should raise error to require `target`" do
+              expect { plugin.register }.to raise_error LogStash::ConfigurationError
+            end
+          end
+        end
+
+        context "`target` is set" do
+          let(:event) { LogStash::Event.new("client" => {"ip" => ip}) }
+          let(:options) { common_options.merge({"source" => "[client][ip]", "target" => target}) }
+
+          context "`target` matches ECS template" do
+            let(:target) { 'host' }
+
+            it "should use `target`" do
+              plugin.register
+              plugin.filter(event)
+              expect( event.get "[#{target}][geo][country_iso_code]" ).to eq 'US'
+            end
+          end
+
+          context "`target` in canonical field reference syntax matches ECS template" do
+            let(:target) { '[host]' }
+
+            it "should normalize and use `target`" do
+              expect(plugin.logger).to receive(:warn).never
+              plugin.register
+              plugin.filter(event)
+              expect( event.get "[host][geo][country_iso_code]" ).to eq 'US'
+            end
+          end
+
+          context "`target` does not match ECS template" do
+            let(:target) { 'host_ip' }
+
+            it "should use `target` with warning" do
+              expect(plugin.logger).to receive(:warn)
```

Comment:
nit: assert the contents of the warning

```suggestion
              expect(plugin.logger).to receive(:warn).with(/ECS expect `target`/)
```